src/lib/utils/DY.Lib.Printing.fst

```diff
@@ -0,0 +1,242 @@
+    | None -> (
+      let str = state_to_str sess_bytes in
+      match str with
+      | Some s -> s
+      | None -> (bytes_to_string sess_bytes)
+    )
+  )
+
+
+(*** Functions to Print the Trace ***)
+
+val option_to_str: (bytes -> option string) -> bytes -> string
+let option_to_str parse_fn elem =
+  let parsed = parse_fn elem in
+  match parsed with
+  | Some str -> str
+  | None -> bytes_to_string elem // Parse bytes with the default method as a fallback
+
+val trace_event_to_string: 
+  trace_event -> nat -> 
+  (bytes -> option string) -> (bytes -> option string) -> (bytes -> option string) -> 
+  string
+let trace_event_to_string tr_event i message_to_str state_to_str event_to_str =
+  match tr_event with
+  | MsgSent msg -> (
+    let msg_str = option_to_str message_to_str msg in
+    Printf.sprintf "{\"TraceID\": %d, \"Type\": \"Message\", \"Content\": \"%s\"}\n"
+      i msg_str
+  )
+  | RandGen usg lab len -> (
+    Printf.sprintf "{\"TraceID\": %d, \"Type\": \"Nonce\", \"Usage\": \"%s\", \"Label\": \"%s\"}\n" 
+    i (usage_to_string usg) (label_to_string lab)
+  )
+  | Corrupt prin sess_id -> ""
+  | SetState prin sess_id full_content -> (
+    let content = get_state_content full_content in
+    let content_str = state_to_str_helper state_to_str content in
+    Printf.sprintf "{\"TraceID\": %d, \"Type\": \"Session\", \"SessionID\": %d, \"Principal\": \"%s\", \"Content\": \"%s\"}\n"
+      i sess_id prin content_str
+  )
+  | Event prin tag content -> (
+    let content_str = option_to_str event_to_str content in
+    Printf.sprintf "{\"TraceID\": %d, \"Type\": \"Event\", \"Principal\": \"%s\", \"Tag\": \"%s\", \"Content\": \"%s\"}\n" 
+      i prin tag content_str
+  )
+
+/// Helper function for `trace_to_string` to avoid calling `length` for each trace event,
+/// which would lead to quadratic complexity.
+
+val trace_to_string_helper:
+  (tr:trace) -> (i:nat{i = DY.Core.Trace.Type.length tr}) ->
+  (bytes -> option string) -> (bytes -> option string) -> (bytes -> option string) ->
+   string
+let rec trace_to_string_helper tr i message_to_str state_to_str event_to_str =
+  match tr with
+  | Nil -> ""
+  | Snoc ptr ev -> (
+      trace_to_string_helper ptr (i-1) message_to_str state_to_str event_to_str ^ 
+      trace_event_to_string ev i message_to_str state_to_str event_to_str
+  )
+
+(*** Functions for Users ***)
+
+/// This method can be used to print the trace of
+/// an example protocol run to the console.
+/// The output can be parsed and converted to a
+/// Plantuml sequence diagram with the following
+/// experimental Python script: https://github.com/fabian-hk/dolev-yao-star-output-parser
+///
+/// Usage:
+/// let* tr = get_trace in
+/// let _ = IO.debug_print_string (
+///           trace_to_string tr default_message_to_str default_state_to_str default_event_to_str
+///         ) in
+
+val trace_to_string: 
+  trace -> 
+  (bytes -> option string) -> (bytes -> option string) -> (bytes -> option string) ->
```

Comment:
Maybe it would be more explicit to use a record here, something like

```fstar
type trace_to_string_printers = {
  message_to_string: bytes -> option string;
  state_to_string: bytes -> option string;
  event_to_string: bytes -> option string;
}
```
Or following ideas in the previous comment:
```fstar
type trace_to_string_printers = {
  message_to_string: bytes -> option string;
  state_to_string: list (string & (bytes -> option string));
  event_to_string: list (string & (bytes -> option string));
}
```